[Navigation editor] Save menu items using the REST API

adamziel opened the pull request:
# Description

This PR migrates navigation saving from `admin-ajax.php` to REST API v2. It ended up being pretty large PR, we may or may not need to split it into a few smaller ones.

# Known issues

**Partial failure is very much possible**. A lot of validation happens in [`prepare_item_for_database`](https://github.com/WordPress/gutenberg/blob/818018d922f865b8ef9dae85ce28f35284ef7bc2/lib/class-wp-rest-menu-items-controller.php#L380) so we can't rely on `require-all-validate` feature of batch requests. The more validation we move to `validate_callback`, the more reliable this will be.

I don't think there is a good way of communicating the partial failure to the user so we should minimize the risk as much as possible. Here's how it looks like for now:

<img width="537" alt="Zrzut ekranu 2021-09-9 o 17 39 47" src="https://user-images.githubusercontent.com/205419/132717814-9a4a5241-cbf2-4cac-b6af-631be62e4402.png">

# Other considerations

Editor is in a dirty state after loading the page. This is not specific to this PR – trunk behaves the same.

**What about using `<EntityProvider />`? Would it make the code simpler?**

We could get rid of the "stub post", that's for sure. About batch saving – we'd still need to compute the diff between what's persisted and what's in memory. **I think** that `<EntityProvider />` would just call `blockToMenuItem` earlier than we do now, but we'd still need to compute a list of batch tasks and check for errors.

**Other changes:**
* `serializeProcessing` was replaced by locks from `core-data`
* Generator actions were replaced by async functions
* Most controls are gone as they were no longer used

Related to https://github.com/WordPress/gutenberg/issues/24907

# Test plan

Go to the experimental navigation screen, play with it, and keep saving the data. Expect to see frequent API errors, but no JS errors.

adamziel committed:
First stab at saving menu items using the REST API

jsnajdr commented:
> I'd like to use thunks here but we'd need to figure out how to make registry selectors

Every thunk receives `registry` as a parameter, so you can write code like this:
```js
const thunk = () => ( { registry } ) => {
  const error = registry.select( 'core' ).getLastEntitySaveError( 'root', 'menu', menuId );
  /* ... */
}
```

adamziel commented:
> Every thunk receives registry as a parameter, so you can write code like this:

Uuu that's nice and I totally missed that, thank you!

talldan commented on packages/edit-navigation/src/store/actions.js:
```diff
@@ -184,34 +189,94 @@ function mapMenuItemsByClientId( menuItems, clientIdsByMenuId ) {
+
+		batchTasks.push( ( { saveEditedEntityRecord } ) =>
+			saveEditedEntityRecord( 'root', 'menuItem', menuItem.id )
+		);
```
It'd be cool if there were an API like `const batch = createBatchSaveForEditedEntityRecords( 'root', 'menuItem' )`

adamziel replied:
👍  I like it!

adamziel replied:
I think this PR comes pretty close now.

adamziel committed:
Migrate batch save to async thunks using REST API calls

adamziel commented on packages/edit-navigation/src/store/actions.js:
```diff
@@ -49,125 +38,141 @@ export function setSelectedMenuId( menuId ) {
+} ) => {
 	const menuId = post.meta.menuId;
+	// @TODO: extract locks to a separate package?
+	const lock = await registry
```
Serial processing was just locks

adamziel commented:
Still pretty draft-y, but I like how thunks + locks = -500 lines so far

adamziel committed:
simplify the logic

adamziel committed:
Simplify computeDeletedMenuItemIds

adamziel committed:
Use more generic variable names

adamziel committed:
Use more consistent variable names

adamziel committed:
Update the data flow to ensure API receives appropriate parameters names

adamziel committed:
Alter the order of arguments in blockToMenuItem

adamziel committed:
Move transformations to transform.js

adamziel committed:
Remove dev comment

adamziel committed:
Operate on lists and sets, not on trees

adamziel commented:
Not quite there yet, but we're getting close. `createBatchSave()` looks pretty generic now, I wonder if we could apply the same "here's a diff, please persist it" structure for widgets. Maybe it would even be a good candidate for core-data? `computeNewMenuItems` is also pretty generic if we parametrize with `blockToMenuItem`. FYI @jsnajdr @noisysocks

adamziel committed:
Extract batchSaveMenuItems

adamziel committed:
Inline blockIdToOldEntityRecord

spacedmonkey committed:
Remove tests

adamziel committed:
Ensure consistency between stored menu items and the ones recovered from the block attributes

adamziel committed:
Add error checking

adamziel committed:
Add generic batchSaveDiff function

adamziel committed:
Preprocess menu data before sending it over to the API

adamziel committed:
Adjust unit tests

adamziel committed:
Remove dev artifacts

adamziel commented on packages/core-data/src/entities.js:
```diff
@@ -133,6 +135,27 @@ export const kinds = [
+		newEdits.type = 'custom';
+	}
+	return newEdits;
+}
```
This ensures that all the menu item fields are always sent in the PUT request – ideally we wouldn't need it. I have a hunch that solving the `invalid_menu_order` situation could also help us here.

talldan replied:
Have you looked at https://github.com/WordPress/gutenberg/pull/34672?

spacedmonkey replied:
It was agreed to just remove the validation. Which is why I worked on #34672. Once that PR is merged, rebase and you should be good to go.

adamziel commented:
Aside of the API problems, I think this is in a pretty good shape. Let's discuss! CCing previously involved folks @noisysocks @talldan @draganescu @TimothyBJacobs @spacedmonkey

adamziel commented on packages/edit-navigation/src/store/test/actions.js:
```diff
@@ -151,289 +57,53 @@ describe( 'createMissingMenuItems', () => {
 	} );
 } );
 
-describe( 'saveNavigationPost', () => {
```
I removed these `saveNavigationPost` tests for now as they are customizer-specific. Let's discuss the ideas in this PR and then double down with more relevant tests.

adamziel replied:
We could use unit tests for the async REST API-based version, but I think it's okay to follow-up with that to avoid inflating this PR more than it already it.

getdave commented on packages/edit-navigation/src/store/transform.js:
```diff
@@ -84,41 +212,20 @@ function mapMenuItemsToBlocks( menuItems ) {
- * For more documentation on the individual fields present on a menu item please see:
- * https://core.trac.wordpress.org/browser/tags/5.7.1/src/wp-includes/nav-menu.php#L789
- *
- * Changes made here should also be mirrored in packages/edit-navigation/src/store/utils.js.
```
Just noting this. Much of this code is mirrored in the block so we should consider that.

adamziel replied:
Block library seems to only use it to create a placeholder. Perhaps we could remove it from there? Do the blocks need to know about the underlying data model?

adamziel replied:
Tracked in https://github.com/WordPress/gutenberg/issues/35037

talldan commented on packages/core-data/src/entities.js:
```diff
@@ -133,6 +135,27 @@ export const kinds = [
+		...persistedRecord,
+		...edits,
+	};
+	if ( Array.isArray( newEdits.menus ) ) {
+		newEdits.menus = newEdits.menus[ 0 ];
+	}
+	if ( newEdits.type === '' ) {
+		newEdits.type = 'custom';
+	}
```
I don't fully understand why these are needed, the `type` one feels like a bug. Would be good to add comments above these if statements.

adamziel replied:
Thanks for marking that. I want to revisit that part after the REST API PR lands.

spacedmonkey replied:
We can set a default value in the REST API, if that is helpful...

talldan commented on packages/edit-navigation/src/store/actions.js:
```diff
@@ -49,169 +38,307 @@ export function setSelectedMenuId( menuId ) {
+				entityIdToBlockId[ entityRecord.id ],
+				entityRecord,
+			] )
+			.filter( ( [ blockId ] ) => blockId )
```
Could the whole thing be done as a single `reduce` into an accumulated object? I feel like that might be more readable.

Although I know some find reduce hard to understand. Extra code comments would be useful either way.

adamziel replied:
Yeah I looked at after a day and had a hard time too. I replaced it with a good, old for loop. I didn't add any comments for now because I couldn't come up with anything better than "map block ids to their corresponding entity records" and the variable name is `blockIdToOldEntityRecord` – what do you think?

kevin940726 replied:
Huh I actually don't think this is hard to read to be honest. I'd imagine `reduce` to be worse, but maybe that's just me. Though I agree for-loop would be better in general.

talldan replied:
A loop seems like a good compromise, thanks 👍

talldan commented on packages/edit-navigation/src/store/actions.js:
```diff
@@ -49,169 +38,307 @@ export function setSelectedMenuId( menuId ) {
-} );
+};
+
+const computeNewMenuItems = ( post, oldMenuItems ) => async ( {
```
The word 'new' does make me think that this will return menu items that have been created in the editor since the last save, but reading the code I don't think that's the case.

I think it's more that the diff is 'saved menu items' vs 'current menu items'. On the other hand, I guess some might consider 'current' to mean 'currently saved' 🥲 

So possibly some naming here that can be improved.

adamziel replied:
Good point! After some thinking, I think "desired" might be a better word here.

talldan reviewed (commented):
This is an amazing amount of work, great job!

This all makes sense to me, though I'd say that the code did take me a while to read through and fully understand, so I think there's a lot that can be done to improve both the self-documenting nature of the code and the explicit documentation (doc blocks and inline comments).

spacedmonkey commented:
I did some testing and when I tried to save a empty menu item ( no label / title ). I got a REST API error. The editor should have this state. 

<img width="571" alt="Screenshot 2021-09-10 at 10 45 45" src="https://user-images.githubusercontent.com/237508/132835136-4c21f679-5b3e-4369-9548-bb24b62afb41.png">

adamziel committed:
Validate if menu_order is set and >= 1

adamziel committed:
Replace map/filter with a for loop for readability

adamziel committed:
Rename "new" to Desired" to avoid abiguity

adamziel committed:
Rename tasks and diffs to changesets

adamziel commented:
> I got a REST API error. The editor should have this state.

I wonder what we should do there 🤔 I don't think we have a way of representing validation errors in the editor – we just don't allow the editor to represent an invalid state in the first place. Perhaps we could just allow an empty title in the API? After all, why not? A different idea would be to display a placeholder in the editor and use it for saving, but I like that less.

adamziel commented:
> This all makes sense to me, though I'd say that the code did take me a while to read through and fully understand, so I think there's a lot that can be done to improve both the self-documenting nature of the code and the explicit documentation (doc blocks and inline comments).

You got it @talldan, I just improved some of the terms we use. I'll continue next week and also fill in the remaining blanks with documentation.

spacedmonkey commented:
> I wonder what we should do there 🤔 I don't think we have a way of representing validation errors in the editor – we just don't allow the editor to represent an invalid state in the first place. Perhaps we could just allow an empty title in the API? After all, why not? A different idea would be to display a placeholder in the editor and use it for saving, but I like that less.

Look at the current editor. 

https://user-images.githubusercontent.com/237508/132948201-d9d6df1b-b756-485a-8cee-71f837cd5050.mov

It defaults to "Menu item" as a default title. If the title is empty is saved, then this menu item is remove. I think just copy this behaviour.

noisysocks commented on packages/edit-navigation/src/store/actions.js:
```diff
@@ -49,169 +38,311 @@ export function setSelectedMenuId( menuId ) {
-		stack.push( ...block.innerBlocks );
+
+		dispatch( {
+			type: 'SET_MENU_ITEM_TO_CLIENT_ID_MAPPING',
```
We should probably switch to storing the menu item ID in a block attribute (`__internalMenuItemId`) as client IDs aren't stable. Same as https://github.com/WordPress/gutenberg/pull/28379. Later, though 🙂

adamziel replied:
Yeah let's follow up with another PR.

karmatosed commented:
Just popping up my head and would like to gently nudge us to use existing notice methods, at least for the baseline. I see at the start, for example, in this issue snack bars recommended, for things such as ‘not saving’ this is a little too brief I feel for a snack bar.

> It defaults to "Menu item" as a default title. If the title is empty is saved, then this menu item is remove. I think just copy this behaviour.

This works for me, and whilst you could see it not as totally ideal, if possible, where the interface is the same as the old starting with the baseline makes sense. We can absolutely improve and iterate up, but starting there is a great springboard.

jsnajdr commented on packages/edit-navigation/src/store/actions.js:
```diff
@@ -49,169 +38,311 @@ export function setSelectedMenuId( menuId ) {
+			mapping: menuItemIdToBlockId,
+		} );
+	} finally {
+		await registry.dispatch( 'core' ).__unstableReleaseStoreLock( lock );
```
unlocking is sync.

jsnajdr commented on packages/edit-navigation/src/store/controls.js:
```diff
@@ -127,38 +39,6 @@ export function dispatch( registryName, actionName, ...args ) {
-		}
-	),
-
 	DISPATCH: createRegistryControl(
```
The `dispatch` and `resolveMenuItems` controls could also easily be removed after the store resolvers are thunkified.

adamziel replied:
Tracked in https://github.com/WordPress/gutenberg/pull/35044

jsnajdr commented on packages/edit-navigation/src/store/actions.js:
```diff
@@ -49,169 +38,311 @@ export function setSelectedMenuId( menuId ) {
+
+const getEntityRecordIdToBlockIdMapping = ( postId ) => async ( {
+	registry,
+} ) => registry.stores[ STORE_NAME ].store.getState().mapping[ postId ] || {};
```
This could be converted to a selector on the store's state, completely sync.

adamziel replied:
It could, but we didn't want this to be a public API. It will likely be gone in the next PR anyway when we move to using internal attributes

jsnajdr replied:
In a thunk, you can use the `select` function to run a private selector. You've already dispatched private actions with `dispatch( { type: FOO } )`, and this is very similar:
```js
const getEntityRecordIdToBlockIdMapping = ( postId ) => ( state ) => state.mapping[ postId ] || {};

const getDesiredMenuItems = ( post, oldMenuItems ) => ( { select } ) => {
  const entityIdToBlockId = select( getEntityRecordIdToBlockIdMapping( post.id ) );
  ...
}
```

jsnajdr commented on packages/edit-navigation/src/store/actions.js:
```diff
@@ -49,169 +38,311 @@ export function setSelectedMenuId( menuId ) {
+const resolveSelectMenuItems = ( menuId ) => async ( { registry } ) =>
+	await registry
+		.resolveSelect( 'core' )
+		.getMenuItems( { menus: menuId, per_page: -1 } );
```
This is also not really an action. What about inlining the function into the two places that call it?

jsnajdr commented on packages/edit-navigation/src/store/actions.js:
```diff
@@ -49,169 +38,311 @@ export function setSelectedMenuId( menuId ) {
+		);
+
+	return menuItem;
+};
```
The `createPlaceholderMenuItem` action will create a new menu item on the server, then download the full list from the server, and store the list+new item in state? The way how it's implemented looks unreliable and racy. The `resolveSelectMenuItems` call might download the full list, with the new item already present, and then add a duplicate copy of the new item.

adamziel replied:
Great spot!

jsnajdr commented on packages/edit-navigation/src/store/actions.js:
```diff
@@ -49,169 +38,311 @@ export function setSelectedMenuId( menuId ) {
+			resolveSelectMenuItems( post.meta.menuId )
+		);
+		const desiredMenuItems = await dispatch(
+			getDesiredMenuItems( post, oldMenuItems )
```
`getDesiredMenuItems` is a sync selector.

jsnajdr commented on packages/edit-navigation/src/store/actions.js:
```diff
@@ -49,169 +38,311 @@ export function setSelectedMenuId( menuId ) {
-			{
+		await registry
+			.dispatch( noticesStore )
+			.createSuccessNotice( __( 'Navigation saved.' ), {
```
`createSuccessNotice` is sync.

jsnajdr commented on packages/edit-navigation/src/store/actions.js:
```diff
@@ -49,169 +38,311 @@ export function setSelectedMenuId( menuId ) {
+			.dispatch( noticesStore )
+			.createErrorNotice( errorMessage, {
+				type: 'snackbar',
+			} );
```
Also sync.

jsnajdr commented on packages/edit-navigation/src/store/actions.js:
```diff
@@ -49,169 +38,311 @@ export function setSelectedMenuId( menuId ) {
+				type: 'snackbar',
+			} );
+	} finally {
+		await registry.dispatch( 'core' ).__unstableReleaseStoreLock( lock );
```
sync

jsnajdr commented on packages/edit-navigation/src/store/actions.js:
```diff
@@ -49,169 +38,311 @@ export function setSelectedMenuId( menuId ) {
+	desiredEntityRecords
+) => async ( { dispatch, registry } ) => {
+	const changeset = await dispatch(
+		computeChangeset( kind, type, oldEntityRecords, desiredEntityRecords )
```
`computeChangeset` is a sync function that computes a return value and also has side-effects: applying edits to entity in the state.

jsnajdr commented on packages/edit-navigation/src/store/actions.js:
```diff
@@ -49,169 +38,311 @@ export function setSelectedMenuId( menuId ) {
+		.__experimentalBatch( changeset.map( ( { batchTask } ) => batchTask ) );
+
+	const failures = await dispatch(
+		getFailedChanges( kind, type, changeset, results )
```
`getFailedChanges` is a sync selector that calls `registry.select( 'core' ).getLastEntitySaveError` several times with various arguments and combines the results together.

jsnajdr commented on packages/edit-navigation/src/store/actions.js:
```diff
@@ -49,169 +38,311 @@ export function setSelectedMenuId( menuId ) {
 	const menuId = post.meta.menuId;
+	// @TODO: extract locks to a separate package?
+	const lock = await registry
+		.dispatch( 'core' )
```
Let's `import { store } from '@wordpress/core-data'` and use it here instead of the `'core'` string. The `./selectors.js` module already does that.

adamziel committed:
Remove prePersistMenuItem

adamziel committed:
Don't await releaseStoreLock

adamziel committed:
Inline resolveSelectMenuItems

adamziel committed:
Resolve a list of all menu items prior to creating placeholder menu items.

adamziel committed:
Remove async code from the synchronous getDesiredMenuItems selector

adamziel committed:
Don't await createErrorNotice and createSuccessNotice

adamziel committed:
Rename computeChangeset to prepareChangeset and remove awaits from it and getFailedChanges

adamziel committed:
Use store object to refer to core-data store, not a string identifier

adamziel committed:
Set placeholder menu order to 1

adamziel committed:
Don't create placeholder menu items for non-link blocks

adamziel committed:
Remove menu order validation

adamziel committed:
Remove menu id validation

adamziel committed:
Restore rest_invalid_param validation error code

adamziel commented:
I addressed most of the feedback on this PR, let's punt these points to the next PR:

* Use internal block attributes instead of mapping client ids to block ids (from @noisysocks )
* Have "Menu Item" as the default title (from @spacedmonkey )
* Migrate the other store resolvers to thunks (from @jsnajdr )

I will add the missing documentation, adjust any failing tests, and I think we'll be good to merge.

adamziel committed:
Fix unit tests

adamziel committed:
Add comments blocks

adamziel commented:
The doc blocks are in, we may be good to land this one! CC @talldan @kevin940726 @noisysocks

spacedmonkey commented on packages/edit-navigation/src/store/test/resolvers.js:
```diff
@@ -85,7 +85,7 @@ describe( 'getNavigationPostForMenu', () => {
 				url: 'http://wp.com',
 				menu_order: 1,
-				menus: [ 1 ],
+				menus: 1,
```
👍

noisysocks commented:
It looks like this causes submenus to no longer save. It works in `trunk`.

https://user-images.githubusercontent.com/612155/133354713-b3227c6e-b7a0-4cc7-b028-4c06731be671.mp4

noisysocks commented on lib/class-wp-rest-menu-items-controller.php:
```diff
@@ -391,6 +391,7 @@ protected function prepare_item_for_database( $request ) {
 				'menu-item-object'      => $menu_item_obj->object,
 				'menu-item-parent-id'   => $menu_item_obj->menu_item_parent,
 				'menu-item-position'    => $position,
+				'menu-item-type'        => $menu_item_obj->type,
```
What's this change? Can we cover it in the menu items controller unit tests?

adamziel replied:
It retains the menu item type if you don't supply one in the request body. Also known as bugfix :-) Sure, I'll add a test for that

spacedmonkey replied:
Seems like a valid change!

noisysocks commented on packages/edit-navigation/src/store/actions.js:
```diff
@@ -49,171 +39,380 @@ export function setSelectedMenuId( menuId ) {
+	registry,
+} ) => {
 	const menuId = post.meta.menuId;
+	// @TODO: extract locks to a separate package?
```
Why? Doesn't `@wordpress/core-data` use locking? If that's the case then may as well have everything all remain in `@wordpress/core-data`.

noisysocks commented on packages/edit-navigation/src/store/actions.js:
```diff
@@ -1,33 +1,23 @@
+import { blockToMenuItem } from './transform';
 
 /**
  * Returns an action object used to select menu.
```
Would be good to update this doc comment. It seems to be describing when this action is dispatched, which is useful, but not what it does.

adamziel replied:
I am confused – that function does return an action object used to select menu. What am I missing?

noisysocks replied:
Oh my bad, the collapsed diff made me think this was the comment of `createMissingMenuItems` 🙂

noisysocks commented on packages/edit-navigation/src/store/actions.js:
```diff
@@ -49,171 +39,380 @@ export function setSelectedMenuId( menuId ) {
+		const blocks = blocksTreeToFlatList( post.blocks[ 0 ].innerBlocks );
+		for ( const { block } of blocks ) {
+			if ( block.name !== 'core/navigation-link' ) {
+				continue;
```
Maybe fail loudly here? In case a plugin has added a non-link block or something weird like that.

adamziel replied:
There are blocks we legitimately want to skip, e.g. a submenu block. Logging a warning seems to be a good trade-off though.

noisysocks replied:
Do we not hide submenu if it's not supported?

talldan replied:
Submenu blocks should be converted to menu items as well btw.

adamziel replied:
Tracked in https://github.com/WordPress/gutenberg/issues/35045

noisysocks commented on packages/edit-navigation/src/store/actions.js:
```diff
@@ -49,171 +39,380 @@ export function setSelectedMenuId( menuId ) {
+ * @param {number} menuId Menu id to embed the placeholder in.
+ * @return {Function} An action creator
+ */
+export const createPlaceholderMenuItem = ( block, menuId ) => async ( {
```
`block` is unused.

adamziel replied:
Great spot! Let's add a lint rule for that.

noisysocks replied:
I don't think we can lint for this because sometimes you need to define a callback function that only uses the second argument passed.

adamziel replied:
Ah yes, that's right

noisysocks commented on packages/edit-navigation/src/store/actions.js:
```diff
@@ -49,171 +39,380 @@ export function setSelectedMenuId( menuId ) {
+			[ ...existingMenuItems, createdMenuItem ],
+			menuItemsQuery( menuId ),
+			false
+		);
```
Can we not simply dispatch a `saveEntityRecord` action to create the new menu item? It should perform a `POST` if there is no `id`. It's pretty cumbersome (and slow?) to do this append and receive operation ourselves.

adamziel replied:
The problem here was with having `menuId` in `query`: we don't want to append placeholder to menu yet, but we want this entity to show up on `select()`.

That being said, maybe we don't need the placeholders at all anymore? They are a customizer relict. Let's see if we can remove them.

adamziel replied:
Alright, I quickly realized we can't. Batch request needs to know all the IDs upfront to set up parent/child relationships.

adamziel replied:
Actually we may be able to do that, we'd just need to send a "preflight" batch to create all the placeholders for us, and then reuse these IDs. Should be simpler than what we're doing here and we will also be able to get rid of locks.

Let's follow up with that change – it will be much simpler without the "client id – block id" mapping thing, and I think we'll have a small debate about using an `__internalMenuItemId` vs `menuItemId` attribute.

spacedmonkey replied:
I think there should be three batch requests to the menu item endpoint. 
1. Create all new menu items. Update react state with all new ids. 
2. Update existing menu ids, using new ids and new positions. 
3. Delete any remove menu items. 

The ordering here is important. Deleting last makes sense, as if any the other request fail, then at least you try and restore data. Create items first is needed as you need the ids. 

Breaking into three requests as the other benefit, of making extremely unlike that multiple create/edit/deletes batched requests. Doing it this way is also more restful.

noisysocks replied:
All good ideas 👍

talldan commented:
Yep, the changes from https://github.com/WordPress/gutenberg/pull/34281 will have to be replicated in this branch. Where any mapping from blocks to menu items or menu items to blocks take place, the submenu block will have to be taken into account.

noisysocks commented on packages/edit-navigation/src/store/actions.js:
```diff
@@ -49,171 +39,380 @@ export function setSelectedMenuId( menuId ) {
+	oldEntityRecords,
+	desiredEntityRecords
+) => async ( { dispatch, registry } ) => {
+	const changeset = dispatch(
```
Why have this intermediary `changeset` array? Can't `batchSaveChanges` build the list of tasks given to `__experimentalBatch` itself?

noisysocks replied:
Studying it more closely, I suppose that it's to allow `batchSaveChanges` to be written in a generic way such that it could live in `@wordpress/core-data`. I get that, but... YAGNI? 🙂

In my experience it's better to introduce a new abstraction too late than too early as doing the latter can lead to codebases that have a lot of indirection.

Up to you. I'd feel better about this approach if `saveWidgetArea` used it too.

adamziel replied:
Being generic is nice, but the real purpose is to encapsulate `getFailedChanges` logic – it needs to reason differently about saves and deletes.

In my perfect world, we wouldn't need to call `getLastEntitySaveError` or anything like that because `__experimentalBatch` would throw something actionable on failure, but that's beyond the scope of this PR. Once we get there, though, we could simplify this even more.

> Up to you. I'd feel better about this approach if saveWidgetArea used it too.

This is what I'm building up to! The truth is that we could use the same structure in any editor where you manipulate entities through blocks.

noisysocks replied:
I trust you!

noisysocks commented on packages/edit-navigation/src/store/actions.js:
```diff
@@ -49,171 +39,380 @@ export function setSelectedMenuId( menuId ) {
+ * @param {Array} listB List B.
+ * @return {Array} elements of list A that are not in list B.
+ */
+function diff( listA, listB ) {
```
Can we use `_.difference`? WP Admin always includes lodash so making use of it where possible reduces surface area for bugs and number of bytes sent to the browser.

adamziel replied:
good idea!

noisysocks reviewed (commented):
Definitely a big simplification! Nice work 🙂

talldan commented:
> It looks like this causes submenus to no longer save. It works in trunk.

Just checking again now, and it actually seems broken in `trunk`. Must have been a regression in the last day, pretty sure it was working yesterday.

adamziel committed:
Remove "@TODO: extract locks to a separate package?"

adamziel committed:
Remove unused block argument from createPlaceholderMenuItem

adamziel committed:
Remove obsolete "block" argument

adamziel committed:
Replace diff() with _.difference()

adamziel commented:
It seems like I broke the tests, I'll fix them up tomorrow.

adamziel committed:
Replace clientId mapping with internal attribute denoting entity record ID

spacedmonkey commented on packages/edit-navigation/src/store/actions.js:
```diff
@@ -49,171 +39,366 @@ export function setSelectedMenuId( menuId ) {
+		path: `/__experimental/menu-items`,
+		method: 'POST',
+		data: {
+			title: 'Placeholder',
+			url: 'Placeholder',
```
I don't love these values. How about these values. 

```suggestion
			title: __( 'Menu item', 'gutenberg' ),
			url: '#placeholder',
```

I like these values better. 
- Url is a valid url now, meaning no validation problems. 
- Title matches the value of the default titleless custom menu item. 
- Title is transable, so if something goes wrong, the title is transable for the user.

If we need placeholder menu items, maybe we could add a field of post meta to the menu item, like is_placeholder and return this in the REST API, thoughts?

adamziel replied:
Good call!

talldan commented on packages/edit-navigation/src/store/actions.js:
```diff
@@ -49,171 +39,366 @@ export function setSelectedMenuId( menuId ) {
 			  )
-			: __( 'Unable to save: An error ocurred.' );
-		yield dispatch( noticesStore, 'createErrorNotice', errorMessage, {
+			: __( 'Unable to save: An error o1curred.' );
```
Small typo seems to have crept in

noisysocks replied:
It's l33tspeak.

talldan replied:
That didn't o1cur to me.

adamziel committed:
Radically simplify batch saving

adamziel committed:
Fix unit tests

adamziel committed:
Translate Menu item title

adamziel committed:
Remove obsolete imports

adamziel commented:
Welcome to Extreme Makeover: ~~Home~~ Gutenberg Edition!

In the most recent episode, we:

* Added separate batch requests for storing placeholders, updating menu items, and deleting them.
* Removed the idea of changesets and embraced throwing exceptions right in the batch tasks.
* Removed `post id -> block id` mapping, and moved all the ids to internal block attributes.
* Removed another 1000 lines of code.
* Fixed edit-navigation unit tests. Let's hope we didn't break end to end ones in the process!

This time we didn't add doc blocks or new tests, but stay tuned for the next episode!

(I know this PR is already huge, I simply couldn't resist these changes once I saw how these data structures make me just shuffle glue code left and right)

adamziel commented on lib/class-wp-rest-menu-items-controller.php:
```diff
@@ -671,7 +671,7 @@ public function prepare_item_for_response( $post, $request ) {
 		if ( rest_is_field_included( 'parent', $fields ) ) {
 			// Same as post_parent, expose as integer.
-			$data['parent'] = (int) $menu_item->post_parent;
+			$data['parent'] = (int) $menu_item->menu_item_parent;
```
Without this change, the API always returns parent ID = 0 even though it sets it correctly on PUT. Let's revisit the naming of these fields in another PR.

spacedmonkey replied:
This is fixed in trunk. I recommend rebasing. We fixed a couple of things on the rest api.

spacedmonkey replied:
See https://github.com/WordPress/gutenberg/pull/34835

spacedmonkey commented:
I wrote a little script that creates a new menu and creates 500 items. This script looks like this. 
```
function create_mock_menu( ) {
	$rand = rand();
	for ($x = 0; $x <= 500; $x++) {
		$test = wp_insert_post( [ 'post_type' => 'nav_menu_item', 'post_status' => 'publish', 'menu_order' => $x,'post_title' => 'Title '.$x ] );
		update_post_meta( $test, '_menu_item_type', 'custom' );
		update_post_meta( $test, '_menu_item_url', '#' );

		wp_set_post_terms( $test, 'Test menu '.$rand, 'nav_menu' );
	}
}
```

My testing, there seem to be some performance issues with saving 500 items. See this video. 

https://user-images.githubusercontent.com/237508/133945695-90bb5931-7474-4dc2-85f0-26cc2aa0aab7.mov

Issues, I spotted. 
1. Loading is very slow and does start render when items are coming in, but waits for all items to load. 
2. When pressing save, the save button should be disabled. 
3. When pressing save, a spinner or other something should show the user that menu is saving. 
4. There are lots of get rest api requests. One for each item. Instead of loading each item, just request all items in this menu again. This would be 5 requests of 100 and not 500 requests. 500 requests this close together might take the server down or trigger firewalls. 


Also, in testing, I managed to also get this error message. 
<img width="292" alt="Screenshot 2021-09-19 at 23 36 25" src="https://user-images.githubusercontent.com/237508/133946439-26dcb6de-92ab-4e65-bc7f-131ea3e6dfa8.png">

Through I couldn't workout how I did this.

noisysocks commented on packages/edit-navigation/src/store/actions.js:
```diff
@@ -163,57 +99,130 @@ export const saveNavigationPost = serializeProcessing( function* ( post ) {
+				url: '#placeholder',
+				menu_order: 1,
+			} );
+			block.attributes.__internalRecordId = record.id;
```
I wouldn't expect that mutating a block like this should work. Don't we need to create a copy and persist it using `receiveBlocks()`?

adamziel replied:
This is just to mutate the list, we don't want to actually update the post just yet. This is the last step in the process after all the requests are made. I took that snippet from the widgets editor here:

https://github.com/WordPress/gutenberg/blob/2449a3ac3837721a0db58bed2c5304820873661c/packages/edit-widgets/src/store/actions.js#L223-L225

adamziel replied:
After some more thinking, I replaced mutating block attributes that obscured the data flow, with immutability. Now we explicitly keep track of the updated block objects and update the post with them.

noisysocks commented on packages/edit-navigation/src/store/actions.js:
```diff
@@ -42,119 +30,67 @@ export function setSelectedMenuId( menuId ) {
-		if ( ! batchSaveResponse.success ) {
-			throw new Error( batchSaveResponse.data.message );
-		}
+		await dispatch( batchInsertPlaceholderMenuItems( annotatedBlocks ) );
```
It might be nice to add a comment here or somewhere explaining why we need to break up this operation into seperate requests.

adamziel replied:
Good call, added

noisysocks commented on packages/edit-navigation/src/store/actions.js:
```diff
@@ -163,57 +99,130 @@ export const saveNavigationPost = serializeProcessing( function* ( post ) {
+				if ( failure ) {
+					throw new Error( failure );
+				}
+				block.attributes = menuItemToBlockAttributes(
```
Same again here regarding mutation 🙂

adamziel replied:
There are no more mutations :-) see https://github.com/WordPress/gutenberg/pull/34541#discussion_r712192778

noisysocks commented on packages/edit-navigation/src/store/actions.js:
```diff
@@ -42,119 +30,67 @@ export function setSelectedMenuId( menuId ) {
+				getRecordIdFromBlock( block )
+			)
 		);
+		await dispatch( batchDeleteMenuItems( deletedIds ) );
```
Why can't deletion happen in the same batch as updating?

spacedmonkey replied:
See https://github.com/WordPress/gutenberg/pull/34541#discussion_r709534744

It is just a little cleaner to delete items after we know that create and update requests have completely successfully.

noisysocks reviewed (commented):
I really like the latest round of changes! ❤️

adamziel committed:
Switch to using immutable data structures

adamziel committed:
Add doc blocks

adamziel committed:
Add a test for default 'menu-item-type'

adamziel committed:
Restore newline

adamziel commented on packages/edit-navigation/src/store/transform.js:
```diff
@@ -31,12 +163,10 @@ export default function menuItemsToBlocks( menuItems ) {
 	const sortedItems = sortBy( menuItems, 'menu_order' );
 
-	const innerBlocks = sortedItems.map( ( menuItem ) => {
+	const blocks = sortedItems.map( ( menuItem ) => {
```
Noting for later: This doesn't match nested blocks with their related menu item ids – let's figure it out in another PR.

talldan replied:
👍 Sounds good. Would be great to see an issue for it.

adamziel replied:
Tracked in https://github.com/WordPress/gutenberg/issues/35047

adamziel commented:
@spacedmonkey all good notes. I don't think we have to block this PR on that, let's save them for a follow-up one. The proliferation of GET requests could be something with core-data and resolvers.

adamziel commented:
Alright, I think we're good now with all the tests passing, doc blocks in place, and we have one more php unit test. This is not to say we're free of problems, quite the opposite. But with this PR things are in better shape than without it. I'd like to stop tweaking and start merging and iterating. Can anyone spot any blockers, or are we okay to move on?

adamziel committed:
Lint

TimothyBJacobs commented on phpunit/class-rest-nav-menu-items-controller-test.php:
```diff
@@ -544,6 +544,42 @@ public function test_update_item() {
+		$new_data = $response->get_data();
+
+		// The type shouldn't change just because it was missing from request args.
+		$this->assertEquals( 'block', $new_data['type'] );
```
Test looks good!

talldan commented on packages/edit-navigation/src/store/transform.js:
```diff
@@ -1,12 +1,143 @@
+
+	let attributes;
+
+	if ( block.name === 'core/navigation-link' ) {
```
This still needs to include the submenu block. Otherwise those get saved as `type: 'block'`, but instead they should directly transformable to normal menu items.
```suggestion
	if ( [ 'core/navigation-link', 'core/navigation-submenu' ].includes( block.name ) ) {
```

This change should fix a couple of issues.

First the label for a submenu block somehow ends up being changed to 'Block' after clicking the save button:
<img width="437" alt="Screenshot 2021-09-21 at 12 04 09 pm" src="https://user-images.githubusercontent.com/677833/134109998-7e1ad88f-6059-441c-869a-121b3decffa7.png">

Second, I sometimes see this save error with submenus:
<img width="451" alt="Screenshot 2021-09-21 at 12 03 31 pm" src="https://user-images.githubusercontent.com/677833/134110028-e5acca88-4405-4410-ba25-22eacf512c93.png">

talldan replied:
Seems like `isSupportedBlock` could do this job.

adamziel replied:
Nice, great spot! I updated that

talldan commented on packages/edit-navigation/src/store/actions.js:
```diff
@@ -163,57 +92,248 @@ export const saveNavigationPost = serializeProcessing( function* ( post ) {
+const batchInsertPlaceholderMenuItems = ( navigationBlock ) => async ( {
+	registry,
+} ) => {
+	const blocksWithoutRecordId = blocksTreeToList( navigationBlock )
+		.filter( isSupportedBlock )
+		.filter( ( block ) => ! getRecordIdFromBlock( block ) );
```
I think this could be collapsed to one filter call and it would still look readable. Something like this (I haven't tested this code):
```suggestion
	const blocksWithoutRecordId = blocksTreeToList( navigationBlock )
		.filter( ( block ) => isSupportedBlock( block ) && ! getRecordIdFromBlock( block ) );
```

talldan commented on packages/edit-navigation/src/store/actions.js:
```diff
@@ -163,57 +92,248 @@ export const saveNavigationPost = serializeProcessing( function* ( post ) {
-		body,
+	await registry.dispatch( coreDataStore ).__experimentalBatch( tasks );
+
+	// Throw on failure. @TODO failures should be thrown in core-data
```
Agree with the comment. Also feels like it could be handled as part of the batch, so that we can determine whether the entire batch has an error in one go, rather than iterating through each item.

Fine to handle this in a follow-up.

adamziel replied:
Tracked in https://github.com/WordPress/gutenberg/issues/35048

talldan commented on packages/edit-navigation/src/store/actions.js:
```diff
@@ -163,57 +92,248 @@ export const saveNavigationPost = serializeProcessing( function* ( post ) {
+	return registry
+		.select( coreDataStore )
+		.hasEditsForEntityRecord( 'root', 'menuItem', id );
+};
```
Feels like this function could be inlined if it isn't too messy. When reading the call site, it was hard to determine the kind of value `applyEdits` returns for the filter because name implies a setter instead of a getter. 

Would be good to add docs If you decide to keep it.

talldan reviewed (commented):
Looks good so far, thanks for all the hard work on it. I would be keen to get this merged very soon for testing.

I spotted a couple of issues with submenu blocks, should be an easy fix.

adamziel committed:
Use isBlockSupportedInNav in blockToMenuItem

adamziel committed:
Lint

adamziel committed:
Code style: Collapse two filters into one

adamziel committed:
Inline applyEdits

talldan reviewed (approved):
Thanks Adam. All the issues I noticed seem to be fixed with the latest commits.

I suggest we get this merged to avoid conflicts and then optimise/refine as we go along. Merging will also mean this gets more hands-on testing which is pretty critical at this stage.

A footnote is that ideally this wouldn't be such a huge PR, but I appreciate it required a lot of exploration, and from what I can see the code is cleaner and easier to reason about now.

draganescu commented:
I love that this is in. I wanted to mention that the saving of blocks as menu items is something that started as an experiment, but that in the meantime the understanding of what exactly the block navigation opt in will do has evolved quite significantly. 

So, the focus of optimization and fixing after this 
PR should be on making classic menus work as expected, and less on solving problems around saving blocks in the classic data format. It's highly likely that this path won't be followed in the end.
